File: elementary_data_structures/golang/queue.go
package main

import "fmt"

type Element struct {
	Data int
	Next *Element
}

type Queue struct {
	Size       int
	Head, Tail *Element
}

func (q *Queue) enqueue(data int) {
	q.Size++

	z := &Element{data, nil}

	if q.Head == nil {
		q.Head = z
	} else {
		q.Tail.Next = z
	}
	q.Tail = z
}

func (Q *Queue) dequeue() (int, bool) {
	if Q.Head == nil {
		return 0, false
	}

	Q.Size--

	dequeued := Q.Head
	Q.Head = dequeued.Next

	return dequeued.Data, true
}

func (Q *Queue) printQueue() {
	fmt.Printf("size = %d : ", Q.Size)

	if Q.Head == nil {
		fmt.Println("(empty)")

		return
	}

	for p := Q.Head; p != nil; p = p.Next {
		fmt.Printf("%d", p.Data)
		if p.Next != nil {
			fmt.Print(" -> ")
		} else {
			fmt.Print("\n")
		}
	}
}

func main() {
	Q := &Queue{}

	for {
		fmt.Print("1:enqueue 2:dequeue 3:print > ")
		var op int
		fmt.Scanf("%d", &op)

		switch op {
		case 1:
			fmt.Print("input data > ")
			var data int
			fmt.Scanf("%d", &data)
			Q.enqueue(data)
		case 2:
			dequeued, empty := Q.dequeue()
			if !empty {
				fmt.Println("queue is empty")
			} else {
				fmt.Printf("dequeued data : %d\n", dequeued)
			}
		case 3:
			Q.printQueue()
		default:
			fmt.Println("invalid operation")
		}
	}
}

File: elementary-data-structures/python/queue.py
class QueueEmptyException(Exception):
    pass


class Element:
    __slots__ = ["data", "next"]

    def __init__(self, data):
        self.data = data
        self.next = None


class Queue:
    __slots__ = ["size", "head", "tail"]

    def __init__(self):
        self.size = 0
        self.head = self.tail = None

    def enqueue(self, data):
        self.size += 1

        z = Element(data)

        if not self.head:
            self.head = z
        else:
            self.tail.next = z
        self.tail = z

    def dequeue(self):
        if not self.head:
            raise QueueEmptyException

        self.size -= 1

        dequeued = self.head
        result = dequeued.data
        self.head = dequeued.next

        return result

    def print_queue(self):
        print(f"size = {self.size} : ", end="")

        if not self.head:
            print("(empty)")

            return

        p = self.head
        while p:
            print(p.data, end=" -> " if p.next else "\n")

            p = p.next


if __name__ == "__main__":
    q = Queue()

    while True:
        print("1:enqueue 2:dequeue 3:print > ", end="")
        if (op := int(input())) == 1:
            print("input data > ", end="")
            data = int(input())

            q.enqueue(data)
        elif op == 2:
            try:
                dequeued = q.dequeue()
                print(f"dequeued data = {dequeued}")
            except QueueEmptyException:
                print("queue is empty")
        elif op == 3:
            q.print_queue()
        else:
            print("invalid operation")
